Preliminary multilingual support

Mes0903 opened the pull request:
mdBook does not natively support multilingual documentation. A common approach is to use a third-party library like mdbook-i18n-helper, which relies on a Gettext-based workflow to enable multilingual support.

However, I believe such a complex system is unnecessary for our case. Therefore, this commit follows the approach suggested in the official mdBook documentation: overriding the `index.hbs` template to add a button that redirects users to the corresponding documentation in another language, thus achieving basic multilingual support.

This commit introduces the following three files under the theme directory:

- `index.hbs`: Adds a redirect button for switching languages.
- `lang-toggle.css`: Defines the styling for the button and the dropdown menu.
- `lang-toggle.js`: Implements the main logic behind the button's behavior.

As a result, two separate books need to be built. However, since mdBook’s `book.toml` does not support multiple `[book]` sections in a single configuration, this commit introduces a configuration template and a CMake script that generates language-specific toml files for each version, which are then built into their respective output folders.

Besides, since translations in other languages are expected to be placed under `src`, all original guide content has been moved into `src/en`.

[1]: https://github.com/google/mdbook-i18n-helpers

Mes0903 commented:
I have no confidence to write the cmake script very well, so I may need some advice here. 

For now, I build the document simply via

```
mkdir build && cd build
cmake ..
```

I dont think this is the correct way but I have no idea what's the correct method.

karnkaul commented on guide/CMakeLists.txt:
This is pretty cool! There's also CMake script mode, which I think this suits better than being a full `CMakeLists.txt` file, which is conventionally used to define build targets. BTW, here's the current [deploy script](https://github.com/cpp-gamedev/learn-vulkan/blob/main/.github/workflows/deploy.yml) for reference.

Mes0903 replied:
I think it's better to update it in this commit, I will try it.

karnkaul commented on guide/.gitignore:
```diff
@@ -1 +1,3 @@
 book
+build/*
+book.toml
```
I'm not sure if removing `book.toml` from source is a good idea, but maybe there's no other clean way if this is now generated by a script...

karnkaul commented on guide/book-template.config:
What does this file do?

Mes0903 replied:
The cmake script will use the `book-template.config` to generate the corresponding `book.toml` for each language. Take `en` as an example, the cmake script will replaced the `@AUTHORS@` and `@LANGUAGE@` in the `book-template.config` by the hardcode setting in cmake script:

```
MakeConfig("en" "Karn Kaul")
```

Now, because the `book.toml` was generated by the script, so I add it into the `.gitignore`. If we want to reserve the `book.toml`, maybe we could simply add `book.toml` for each language manually. Then the script will just doing the build work but not generate any file.

karnkaul reviewed (commented):
Nice! Is there a way to not remove `book.toml` from the source tree? 
When I get some time I can update the CMake code to use [script mode](https://cmake.org/cmake/help/latest/manual/cmake-language.7.html#scripts). Unless you want to try it yourself.

karnkaul commented on guide/build-books.cmake:
Nitpick: filenames are in `snake_case`, better to remain consistent. `build_books.cmake`

Mes0903 commented:
The code now can be tested by

- `cmake -P build.cmake` : `mdbook build` for all language
- `cmake -DSERVE=ON build.cmake` : this will build all translations that registered in the script, and move it into the folder that `mdbook serve` generate, which usually called `book`

It should have no problem for the first one, but for the second command, there is some topic that we can discuss

Take zh-TW as example, the processing is generally as following:

- Build zh-TW document into `.tmp-output/zh-TW`
- Start `mdbook serve` for English
- Wait for `book/` folder to be created by mdbook serve
- Then move `.tmp-output/zh-TW` → `book/zh-TW`
- Clean up `.tmp-output`

Since `mdbook serve` will block the shell, so I open an new shell to run the command:

```
if(WIN32)
  execute_process(
    COMMAND powershell -Command "Start-Process mdbook -ArgumentList 'serve','-d','${CMAKE_CURRENT_SOURCE_DIR}/book'"
    WORKING_DIRECTORY "${CMAKE_CURRENT_SOURCE_DIR}"
  )
else()
  execute_process(
    COMMAND bash -c "mdbook serve -d \"${CMAKE_CURRENT_SOURCE_DIR}/book\" &"
    WORKING_DIRECTORY "${CMAKE_CURRENT_SOURCE_DIR}"
  )  
endif()
```

For now, this work successfully on windows, it come out an new powershell showing the info of `mdbook serve` command, and the original shell keep on the translation folder movement work. But on linux, this didnt work as expected, it didnt open an new shell

The command `mdbook serve` need to take some time to generate the `book` folder, For now, I use the naive way ー sleep, to make sure the translation folder movement would start after the `book` folder was generated:

```
# Move the zh-TW output folder into book/
message(STATUS "Moving the translation folder to final location...")
execute_process(COMMAND ${CMAKE_COMMAND} -E sleep 2)

# Move the translation folder to the book folder
file(MAKE_DIRECTORY "${CMAKE_CURRENT_SOURCE_DIR}/book")
file(RENAME
  "${CMAKE_CURRENT_SOURCE_DIR}/.tmp-output/zh-TW"
  "${CMAKE_CURRENT_SOURCE_DIR}/book/zh-TW"
)
```

I dont think this is an good way, it should be done as sth like await mechanism, is there sth similar in cmake script?

Mes0903 committed:
Preliminary support multilingual

mdBook does not natively support multilingual documentation. A common
approach is to use a third-party library like mdbook-i18n-helper, which
relies on a Gettext-based workflow to enable multilingual support.

However, I believe such a complex system is unnecessary for our case.
Therefore, this commit follows the approach suggested in the official
mdBook documentation: overriding the 'index.hbs' template to add a
button that redirects users to the corresponding documentation in
another language, thus achieving basic multilingual support.

This commit introduces the following three files under the theme
directory:

- 'index.hbs': Adds a redirect button for switching languages.
- 'lang-toggle.css': Defines the styling for the button and the dropdown
  menu.
- 'lang-toggle.js': Implements the main logic behind the button's
  behavior.

As a result, two separate books need to be built. However, since
mdBook’s 'book.toml' does not support multiple '[book]'' sections in a
single configuration.

This commit introduces a CMake script to build different language
versions of the book and place them in their respective directories.
It can be tested by command 'cmake -P build.cmake'.

Due to mdBook's behavior, the 'src' field in 'book.toml' determines the
root URL. Therefore, if the theme directory is placed above the src
directory, mdBook will copy it relative to the 'src' path during the
build, which will cause the theme folder to become unreachable.

Since English is the primary language of this documentation, the default
behavior supports running 'mdbook' command directly from the guide
directory, such as with 'mdbook serve'. However, this will only generate
the English version of the documentation, other languages will not be
built. To build other language versions using the mdbook command
directly, it must manually adjust their respective book.toml files.

All translations are placed under 'guide/translations'. To add a new
language, it need to follow the same folder structure, update the CMake
script, and register the new language in 'theme/index.hbs'.

[1]: https://github.com/google/mdbook-i18n-helpers

karnkaul commented on guide/theme/lang-toggle.css:
We should stick to `snake_case` for filenames: `lang_toggle.css`, `lang_toggle.js`. Maybe `zh-TW` can be an exception if it's standard/convention.

karnkaul commented on guide/translations/.gitignore:
```diff
@@ -0,0 +1 @@
+theme
```
This is generated by `guide/build.cmake`?

Mes0903 committed:
Stick to 'snake_case' for filenames

Mes0903 committed:
Stick to 'snake_case' for file reference

karnkaul reviewed (approved):
Excellent work, congratulations and thanks!

Mes0903 commented:
- After our discussion, we decided not to include the `mdbook serve` flag in the CMake script. If you want to use the `mdbook serve` feature, as mentioned in the commit message, you need to manually build all translations first. Then, after the command `mdbook serve` creates the book folder, move those translation folders into it.

- If you want to add a new language, you must:
  - Add the target language to the lang-list in index.hbs.
  - Modify the supportedLangs array in lang_toggle.js.
  - Finally, register it in build.cmake.

- The filenames of translated documents must match the original (English) filenames. The folder structure for translations looks like this:
	```
	\---zh-TW
	    |   book.toml
	    |
	    \---src
	        |   README.md
	        |   SUMMARY.md
	        |
	        \---(All other folders and filenames must match the primary English folder(`src`) structure)
	```
